modules/vmfcore/src/statistics.cpp

```diff
@@ -0,0 +1,1276 @@
+ * limitations under the License.
+ *
+ */
+#include "vmf/statistics.hpp"
+#include "vmf/metadata.hpp"
+#include "vmf/metadatastream.hpp"
+//#include <cstring>
+//#include <string>
+//#include <memory>
+//#include <sstream>
+//#include <cmath>
+//#include <limits>
+//#include <iomanip>
+
+namespace vmf
+{
+
+// struct StatUpdateMode
+
+#define CASE_VAL2STR( _x ) case _x:  return #_x
+std::string StatUpdateMode::toString( StatUpdateMode::Type val )
+{
+    switch( val )
+    {
+    CASE_VAL2STR( vmf::StatUpdateMode::Disabled );
+    CASE_VAL2STR( vmf::StatUpdateMode::Manual   );
+    CASE_VAL2STR( vmf::StatUpdateMode::OnAdd    );
+    CASE_VAL2STR( vmf::StatUpdateMode::OnTimer  );
+    }
+    VMF_EXCEPTION( vmf::IncorrectParamException, "Unknown enum value: " + to_string( (int)val ));
+}
+#undef CASE_VAL2STR
+
+#define IF_STR2VAL( _x )  if( str == #_x ) return _x
+StatUpdateMode::Type StatUpdateMode::fromString( const std::string& str )
+{
+    IF_STR2VAL( vmf::StatUpdateMode::Disabled );
+    IF_STR2VAL( vmf::StatUpdateMode::Manual   );
+    IF_STR2VAL( vmf::StatUpdateMode::OnAdd    );
+    IF_STR2VAL( vmf::StatUpdateMode::OnTimer  );
+    VMF_EXCEPTION( vmf::IncorrectParamException, "Unknown enum string: " + str );
+}
+#undef IF_STR2VAL
+
+// class StatOpBase: builtin operations
+
+class StatOpMin: public StatOpBase
+{
+public:
+    StatOpMin()
+        {}
+    virtual ~StatOpMin()
+        {}
+
+public:
+    virtual std::string name() const
+        { return StatOpFactory::builtinName( StatOpFactory::BuiltinOp::Min ); }
+    virtual void reset()
+        {
+            std::unique_lock< std::mutex > lock( m_lock );
+            m_value = Variant();
+        }
+    virtual bool handle( StatAction::Type action, const Variant& fieldValue )
+        {
+            std::unique_lock< std::mutex > lock( m_lock );
+            switch( action )
+            {
+            case StatAction::Add:
+                if( m_value.isEmpty() )
+                    m_value = fieldValue;
+                else if( m_value.getType() != fieldValue.getType() )
+                    VMF_EXCEPTION( vmf::TypeCastException, "Type mismatch" );
+                else
+                    switch( m_value.getType() )
+                    {
+                    case Variant::type_integer:
+                        if( fieldValue.get_integer() < m_value.get_integer() )
+                            m_value = fieldValue;
+                        break;
+                    case Variant::type_real:
+                        if( fieldValue.get_real() < m_value.get_real() )
+                            m_value = fieldValue;
+                        break;
+                    default:
+                        VMF_EXCEPTION( vmf::NotImplementedException, "Operation not applicable to this data type" );
+                    }
+                break;
+            case StatAction::Remove:
+                if( m_value.isEmpty() )
```

Comment:
Actually, Remove support isn't necessary - it's just the rudiment of older synchronous implementation. Update/rescan facility (Stat/Worker) calls op->handle() in Add case only, so 'action' argument can be easily removed from ops, as well as return bool status also (now void). This simplifies ops implementation and removes a lot of (never tested) exception throws.